// gdrive-node-uploader/dist/fileRename.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.renameFileInGoogleDrive = void 0;
const googleapis_1 = require("googleapis");
const utils_1 = require("./utils");
const renameFileInGoogleDrive = (_a) => __awaiter(void 0, [_a], void 0, function* ({ fileId, fileUrl, fileName, newName }) {
    if (!fileId && !fileUrl && !fileName) {
        throw new Error('File ID, file URL, or file name must be provided.');
    }
    const auth = yield (0, utils_1.getAuthClient)();
    const drive = googleapis_1.google.drive({ version: 'v3', auth });
    let validatedFileId;
    if (fileId) {
        validatedFileId = fileId;
    }
    else if (fileUrl) {
        validatedFileId = (0, utils_1.validateFileIdOrUrl)({ fileUrl });
    }
    else if (fileName) {
        const res = yield drive.files.list({
            q: `name='${fileName}' and trashed=false`,
            fields: 'files(id)'
        });
        if (res.data.files && res.data.files.length > 0) {
            validatedFileId = res.data.files[0].id;
        }
        else {
            return { status: false, message: 'File not found on Google Drive.' };
        }
    }
    if (!validatedFileId) {
        return { status: false, message: 'Valid file ID could not be determined.' };
    }
    try {
        const updateRes = yield drive.files.update({
            fileId: validatedFileId,
            requestBody: {
                name: newName
            },
            fields: 'id, name, modifiedTime, webViewLink'
        });
        return {
            status: true,
            message: 'File renamed successfully.',
            id: updateRes.data.id,
            name: updateRes.data.name,
            modifiedTime: updateRes.data.modifiedTime,
            downloadUrl: updateRes.data.webViewLink
        };
    }
    catch (error) {
        return { status: false, message: `Error renaming file: ${error.message}` };
    }
});
exports.renameFileInGoogleDrive = renameFileInGoogleDrive;
